src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -63,12 +63,136 @@ private BoundExpression BindIsPatternExpression(IsPatternExpressionSyntax node,
+            }
+            else
+            {
+                return BindUserDefinedRecursivePattern(node, type, operators, operand, operandType, hasErrors, diagnostics);
+            }
+        }
+
+        private BoundPattern BindUserDefinedRecursivePattern(
+            RecursivePatternSyntax node,
+            NamedTypeSymbol type,
+            ImmutableArray<Symbol> operators,
+            BoundExpression operand,
+            TypeSymbol operandType,
+            bool hasErrors,
+            DiagnosticBag diagnostics)
+        {
+            // find the unique operator to use
+            MethodSymbol isOperator = null;
+            var subPatterns = node.PatternList.SubPatterns;
+            var subPatternsCount = subPatterns.Count;
+            foreach (var op in operators)
+            {
+                // check that the operator is well-formed. If not, skip it.
+                var candidate = op as MethodSymbol;
+                if (candidate == (object)null || !WellFormedOperatorIs(candidate)) continue;
+
+                // check that its number of out parameters is the same as the arity of this pattern-matching operation.
+                // PROTOTYPE(patterns): consider how we would support `params` for a variable number of pattern positions, for
+                // PROTOTYPE(patterns): example in a regular expression the library does not know statically how many capture
+                // PROTOTYPE(patterns): groups there are.
+                if (candidate.ParameterCount != subPatternsCount + 1) continue; // not the droid you're looking for
+
+                // ok, we've found a candidate. If we already have a candidate, error.
+                // PROTOTYPE(patterns): it has been suggested that we use something like overload resolution on the shape
+                // PROTOTYPE(patterns): of the patterns to further narrow down which operators we consider. We don't do that.
+                if (isOperator != (object)null)
+                {
+                    // Error: Ambiguous `operator is` declarations found in type {type}
+                    diagnostics.Add(ErrorCode.ERR_AmbiguousOperatorIs, node.Location, type);
+                    hasErrors = true;
+                    break;
+                }
+
+                isOperator = candidate;
+            }
+
+            if (isOperator == (object)null)
+            {
+                // Error: No 'operator is' declaration in {type} was found with {node.PatternList.SubPatterns.Count} out parameters.
+                diagnostics.Add(ErrorCode.ERR_OperatorIsParameterCount, node.Location, type, subPatternsCount);
+                // PROTOTYPE(patterns): should we bind the subpatterns for the semantic model to use,
+                // PROTOTYPE(patterns): even though we don't have an expression to match them against?
+                return new BoundWildcardPattern(node, hasErrors: true);
```

Comment:
We should create a "bound error pattern" to do that. We should also preserve the subpatterns. That is the point of the PROTOTYPE comment just above.